frontend/src/app/user/ProfileEditor.tsx

```diff
@@ -0,0 +1,110 @@
+"use client";
+
+import { useEffect, useMemo, useState } from "react";
+import { UserInfo } from "./ReadOnlyUserCard";
+import { Image, Button, Card, CardHeader, Input, Popover, PopoverContent, PopoverTrigger, Textarea, CardFooter } from "@nextui-org/react";
+import FileInput from "./FileInput";
+
+export default function ProfileEditor({userInfo} : {userInfo: UserInfo}) {
+    const [name, setName] = useState<string>(userInfo.name);
+    const isInvalidName = useMemo(() => {
+        return name == "";
+    }, [name]);
+    const [bio, setBio] = useState<string | undefined>(userInfo.bio);
+    const [photo, setPhoto] = useState<string | undefined>(userInfo.photo);
+    const [newPhoto, setNewPhoto] = useState<File>();
+    
+    // userInfo is constant, do not change for now
+    const hasChanged = useMemo(() => {
+        if (name != userInfo.name) return true;
+        if (bio != userInfo.bio && !(bio == "" && userInfo.bio == undefined)) return true;
+        if (photo != userInfo.photo && !(photo == "" && userInfo.photo == undefined)) return true;
+        return false;
+    }, [name, bio, photo])
+    
+    useEffect(() => {
+        if (newPhoto) {
+            setPhoto(URL.createObjectURL(newPhoto))
+            console.log("Recieved file!");
+        } else {
+            setNewPhoto(undefined);
+            setPhoto(userInfo.photo);
+        }
+    }, [newPhoto, userInfo.photo])
+    
+    const handleDiscard = () => {
+        setName(userInfo.name);
+        setBio(userInfo.bio || "");
+        setPhoto(userInfo.photo);
+        setNewPhoto(undefined);
+    }
+
+    const [profileMessage, setProfileMessage] = useState<string>("");
+    const handleProfileSubmit = async () => {
+        if (name == userInfo.name && bio == userInfo.bio && photo == userInfo.photo) {
+            setProfileMessage("Profile saved!");
+            return;
+        }
+
+        const res = await fetch("https://jsonplaceholder.typicode.com/users/1", {
+            method: "PATCH",
+            body: JSON.stringify({
+                name: name,
+                bio: bio,
+                photo: newPhoto,
+                email: userInfo.email
+            })
+        }).catch((err) => {
+            console.log(err);
+            return {
+                status: 500,
+                ok: false
+            }
+        })
+
+        if (!res.ok) {
+            setProfileMessage("An error occured, please try again later");
+        } else {
+            setProfileMessage("Profile saved!");
+        }
+    }
+    return <div className="flex w-full justify-around gap-12">
+            <div> Your Profile </div>
+            <form className="flex w-1/2 flex-col gap-4">
+                <Card isFooterBlurred className="
+                    border-none
+                    self-center
+                    max-w-[200px] max-h-[200px]"
+                    radius="lg">
+                    <Image
```

Comment:
I will suggest you to use the `Badge` component in NextUI to render profile picture instead of image, you might want to take a look. I will leave it for you to design.

Reply:
But the badges are just small circles on an image?

![image](https://github.com/tryyang2001/CS3213-Frontend-Management-System/assets/55905659/fde2543e-4aa2-4c6c-8e0b-a6f41323fa71)

Are you talking about avatar?
![image](https://github.com/tryyang2001/CS3213-Frontend-Management-System/assets/55905659/da799de4-c404-412c-95e1-b315a5bd2392)

We can go with avatar if it looks nicer

Reply:
Ah yes, I mean avatar, sorry for the mistake.